Categories system for parties

LordMidas opened the pull request:
This allows specifying categories within parties where each category is a party. 

Just like #7 it works. I haven't "beautified" the code or updated most comments yet but it is functional. I'd like to merge this so that going forward we can base off of this to make further improvements.

Darxo commented on mod_dynamic_spawns/classes/party.nut:
```diff
@@ -55,6 +77,17 @@
 			}
 		}
 
+		this.__SubParties = [];
```
I'm confused that you call this field `__SubParties` when it is created purely from categories.
What about `__Categories` instead?

LordMidas replied:
`Categories` is a frontend variable used by "designers" to design parties in code. Whereas in fact they are actual party objects. So the private variable used here is called `SubParties` to make it clear in code that these are party objects and I prefer this.

Darxo commented on mod_dynamic_spawns/classes/party.nut:
```diff
@@ -144,39 +180,118 @@
+
+		// RatioMax is ignored if we do not satisfy the RatioMin yet
+		if (!this.satisfiesRatioMin())
+				return true;
```
weird double indent

Darxo commented on mod_dynamic_spawns/classes/spawn_process.nut:
```diff
@@ -63,115 +80,171 @@
+
+	function getSpawnable( _id )
+	{
+		local spawnables = this.__SubSpawnProcesses.len() == 0 ? this.getParty().getUnitBlocks() : this.getParty().getSubParties();
```
this should make use of the getSpawnables() function defined directly above

Darxo commented on mod_dynamic_spawns/classes/party.nut:
```diff
@@ -144,39 +180,118 @@
+	}
+
+	// Returns true if the ratio of this unitblock would still be below its defined RatioMax if it was to spawn the next unit
+	// _spawnProcess = current spawnprocess reference that includes most important variables
```
this comment is now redundant

Darxo commented on mod_dynamic_spawns/classes/party.nut:
```diff
@@ -144,39 +180,118 @@
+	}
+
+	// Returns true if the ratio of this unitblock would still be above its defined RatioMin if it was to spawn the next unit
+	// _spawnProcess = current spawnprocess reference that includes most important variables
```
this comment is now redundant

Darxo commented on mod_dynamic_spawns/classes/party.nut:
```diff
@@ -144,39 +180,118 @@
 
 	// Returns false if there is a reason why this party is not allowed to spawn anymore. This function only checks for a small subset of all possible reasons.
 	// Checks only against HardMax and HardMin
```
this line is no longer correct and should be removed

Darxo reviewed (approved):
Looks ok

All those changes are somewhat confusing to me.
Could me lack of comments, the fact that I wasn't the one writing them or just the sheer amount of changes taken in all at once.

A small comment or explanation is enough to close the issues I raised during this review if you disagree with my concerns.

LordMidas commented:
> Looks ok
> 
> All those changes are somewhat confusing to me. Could me lack of comments, the fact that I wasn't the one writing them or just the sheer amount of changes taken in all at once.
> 
> A small comment or explanation is enough to close the issues I raised during this review if you disagree with my concerns.

Basically major changes include:
- Adding Categories as a frontend variable for partydefs. This is converted party objects which are stored in `__SubParties` on the inside.
- Spawn processes can now have sub spawn processes (these handle the spawning of subparties aka categories).
- Removed the passing around of `_spawnProcess` as a parameter and each spawnable has a private weak reference to its relevant spawn process.

I've updated the comments and other changes you highlighted.
